### PR TITLE
DS-3406: Sort communities and collections in-memory using a comparator

### PR DESCRIPTION
**Pull request by tomdesair**
Fix for https://jira.duraspace.org/browse/DS-3406 where the sorting of communities and collections happens in-memory and not by Hibernate.

### CODE REVIEW

**Comment by terrywbrady** [comment]
@kshepherd , could you review this one?  The Jira ticket references some of your work.

File: dspace-api/src/test/java/org/dspace/content/comparator/NameAscendingComparatorTest.java
Comment (mwoodiupui) [inline dspace-api/src/test/java/org/dspace/content/comparator/NameAscendingComparatorTest.java]: Should that be "dso1"?  dso2 is not used in this test.

**Review by mwoodiupui — COMMENTED** [review]
Other than one suspected typo, I think this is ready.

**Comment by hardyoyo** [comment]
I've tested this PR in vagrant-dspace and can confirm that collection and community names properly sort in an alphabetical order. @mwoodiupu has some good points about a few things, I'd not have any problem with approving this code after Mark's concerns have been addressed.

**Comment by hardyoyo** [comment]
Based on my testing, I'm going to approve this, however, I will wait for @mwoodiupui to merge it, as he seems to have concerns, still.

**Review by hardyoyo — APPROVED** [review]

File: dspace-api/src/main/java/org/dspace/content/comparator/NameAscendingComparator.java
Comment (pnbecker) [inline dspace-api/src/main/java/org/dspace/content/comparator/NameAscendingComparator.java]: I would suggest to either use `StringUtils.compareIgnoreCases(String, String)` here or to trim the Strings and convert them to either small or capital letters. Currently sorting the community/collections ends up in listing all communities/collection A-Z starting with capital letters first and then listing the communities/collection a-z starting with small letters. I would expect one alphabetical sequence, not two.

**Review by pnbecker — CHANGES_REQUESTED** [review]
I tested this in JSPUI and it works fine. There is just one issue left: currently it sorts communities/collections with a strange respect for small/capital letters.

**Comment by tomdesair** [comment]
Hi @pnbecker and @mwoodiupui 
My last commit addresses your comments: https://github.com/DSpace/DSpace/pull/1684/commits/e358cb84d1c4ade5056f3c35912adfddab660eec

Thanks for the thorough review!

**Review by mwoodiupui — APPROVED** [review]
+1 by inspection.  My concerns are addressed, and overall it looks well.

**Review by hardyoyo — APPROVED** [review]
+1 by review, and the tests were also revised to prove that compareToIgnoreCase does what it's supposed to do.

**Review by hardyoyo — APPROVED** [review]
Pascal's concerns have been addressed.

**Review by pnbecker — APPROVED** [review]
I haven't retested it, but the code looks good to me. Thanks a lot for the change!
+1

**Comment by tdonohue** [comment]
This was cherry-picked to `master` in #1768

**Pull request merged**
